### PR TITLE
[Mobile Payments] Badge tap to pay on eligible devices and stores, until Set up Tap to Pay opened

### LINKED ISSUE #9582: [Mobile Payments] Badge hub menu to show that Tap to Pay exists
## Description
Now we have the Set up Tap to Pay on iPhone flow, we can draw users attention to it, as not everyone will regularly go to the menu.

To do this, we can add a badge to the menu and to guide users through to the flow.

See 88fa1962911a019fea887b11debbe7e4864cd184 for removal of previous badge.

We should show the badge only on devices where it's relevant, and for stores where it's relevant. If a device has already been used for Tap to Pay, we shouldn't badge.

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #9582

## Description

N.B. this follows #9801, but changes the logic and display details substantially. If you'd prefer to review both together, feel free to change the base of this PR to `trunk`.

We've added badges leading to Set up Tap to Pay on iPhone, for eligible devices and stores. This PR resolves the three outstanding logic issues from #9801:

- The badges will be shown on any store or device, regardless of Tap to Pay support or previous activity.
- After removal, the badges will return the next time you open the app.
- Taking a Tap to Pay payment will not remove the badges

Now the badges will only be shown when you are using a supported device, an eligible store, have not previously used Tap to Pay to take a transaction, and will remember when the badges have been dismissed.

## Testing instructions

Using an iPhone XS or newer on iOS 16 or above, and a US-based store with WCPay or the Stripe extension

1. Delete then reinstall the app
2. Open the app
3. Observe that the badges are present on the `Menu` tab, and the `Payments` row in the menu, and the `Set up Tap to Pay on iPhone` row in the IPP menu
4. Tap `Set up Tap to Pay on iPhone`, then tap back
5. Observe that all the badges are gone
6. Force quit then re-open the app
7. Observe the badges are still gone

Using an iPhone XS or newer on iOS 16 or above, and a US-based store with WCPay or the Stripe extension

1. Delete then reinstall the app
2. Open the app
3. Observe that the badges are present on the `Menu` tab, and the `Payments` row in the menu, and the `Set up Tap to Pay on iPhone` row in the IPP menu
4. Tap `Collect Payment` and collect a payment using Tap to Pay on iPhone
5. Observe that all the badges are gone

Using an iPhone X or older, or an iPad, or any device running iOS 15, and a US-based store with WCPay or the Stripe extension

1. Delete then reinstall the app
2. Open the app
3. Observe that the badges are not present on the `Menu` tab, and the `Payments` row in the menu, and the `Set up Tap to Pay on iPhone` row in the IPP menu

Using an iPhone XS or newer on iOS 16 or above, and a CA or GB-based store with WCPay

1. Delete then reinstall the app
2. Open the app
3. Observe that the badges are not present on the `Menu` tab, and the `Payments` row in the menu, and the `Set up Tap to Pay on iPhone` row in the IPP menu

## Screenshots

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift [hunk 1]
```diff
@@ -65,21 +65,24 @@ final class HubMenuViewModel: ObservableObject {
...
+        self.tapToPayBadgePromotionChecker = tapToPayBadgePromotionChecker
         self.stores = stores
         self.featureFlagService = featureFlagService
         self.generalAppSettings = generalAppSettings
         self.switchStoreEnabled = stores.isAuthenticatedWithoutWPCom == false
         observeSiteForUIUpdates()
         observePlanName()
-        listenToNewFeatureBadgeReloadRequired()
-        retrieveShouldShowNewFeatureBadgeOnPaymentsValue()
+        tapToPayBadgePromotionChecker.$shouldShowTapToPayBadges.share().assign(to: &$shouldShowNewFeatureBadgeOnPayments)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:85]: 💯 

Should we wrap this line in a method? Sort of "bind/observe badge changes"?

File: WooCommerce/Classes/ViewModels/MainTabViewModel.swift [hunk 2]
```diff
@@ -93,8 +95,7 @@ final class MainTabViewModel {
         listenToReviewsBadgeReloadRequired()
         retrieveShouldShowReviewsBadgeOnHubMenuTabValue()
 
-        listenToNewFeatureBadgeReloadRequired()
-        retrieveShouldShowNewFeatureBadgeOnHubMenuTabValue()
+        tapToPayBadgePromotionChecker.$shouldShowTapToPayBadges.share().assign(to: &$shouldShowNewFeatureBadgeOnHubMenuTab)
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewModels/MainTabViewModel.swift:98]: Same as `HubMenuViewModel`: Should we wrap this line in a method? Sort of "bind/observe badge changes"?

File: WooCommerce/Classes/ViewModels/MainTabViewModel.swift [hunk 3]
```diff
@@ -38,6 +38,8 @@ final class MainTabViewModel {
 
     private var cancellables = Set<AnyCancellable>()
 
+    let tapToPayBadgePromotionChecker: TapToPayBadgePromotionChecker = TapToPayBadgePromotionChecker()
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewModels/MainTabViewModel.swift:41]: Perhaps for another PR: Should we abstract this class into some `PromotionChecker` protocol to keep it detached from the specific TTP promotion checker? It could make sense also inject the protocol in the initializer for testability and to use this common protocol in the future for other feature announcements.

File: WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuCoordinator.swift [hunk 4]
```diff
@@ -22,26 +22,33 @@ final class HubMenuCoordinator: Coordinator {
 
     private let willPresentReviewDetailsFromPushNotification: () async -> Void
 
+    private let tapToPayBadgePromotionChecker: TapToPayBadgePromotionChecker
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuCoordinator.swift:25]: Same comment as with `MainTabViewModel` regarding abstracting this class into a protocol, as would seem that we're leaking implementation details about the badges into the HubMenu navigation coordinator responsibility.

**Review by iamgabrielma — APPROVED** [review]
Tests well! 

✅ Using an iPhone XS or newer on iOS 16 or above, and a US-based store with WCPay or the Stripe extension
✅ Using an iPhone XS or newer on iOS 16 or above, and a US-based store with WCPay or the Stripe extension
✅ Using an iPhone X or older, or an iPad, or any device running iOS 15, and a US-based store with WCPay or the Stripe extension
✅ Using an iPhone XS or newer on iOS 16 or above, and a CA or GB-based store with WCPay

LGTM! Just shared a couple of comments that can be tackled separately, if needed.

File: WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:85]: Yeah, it is a bit wordy. I'll add this to a list to do when I have a little time.
